# 2.2: Project Planning and Sourcing Data with an API

## Import Libraries

In [1]:
import pandas as pd 
import numpy as np
import os 
import requests
import json
from datetime import datetime

In [2]:
# Create a list with all files in the folder using a list compehension

folderpath = r"C:\Users\valev\CityBike\Data" # make sure to put the 'r' in front
filepaths  = [os.path.join(folderpath, name) for name in os.listdir(folderpath)]

In [3]:
filepaths

['C:\\Users\\valev\\CityBike\\Data\\202201-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202201-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202202-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202202-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202203-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202203-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202204-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202204-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202204-citibike-tripdata_3.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202205-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202205-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202205-citibike-tripdata_3.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202206-citibike-tripdata_1.csv',
 'C:\\Users\\valev\\CityBike\\Data\\202206-citibike-tripdata_2.csv',
 'C:\\Users\\valev\\CityBike\\Data

In [4]:
# Read and concatenate all files simultaneously.   Columns 5 and 7 read as string

df = pd.concat((pd.read_csv(f, dtype={5: str, 7: str}) for f in filepaths), ignore_index = True)


In [5]:
df.shape

(29838806, 13)

In [6]:
df.head()

ride_id  rideable_type               started_at  \
0  BFD29218AB271154  electric_bike  2022-01-21 13:13:43.392   
1  7C953F2FD7BE1302   classic_bike  2022-01-10 11:30:54.162   
2  95893ABD40CED4B8  electric_bike  2022-01-26 10:52:43.096   
3  F853B50772137378   classic_bike  2022-01-03 08:35:48.247   
4  7590ADF834797B4B   classic_bike  2022-01-22 14:14:23.043   

                  ended_at       start_station_name start_station_id  \
0  2022-01-21 13:22:31.463  West End Ave & W 107 St          7650.05   
1  2022-01-10 11:41:43.422             4 Ave & 3 St          4028.04   
2  2022-01-26 11:06:35.227          1 Ave & E 62 St          6753.08   
3  2022-01-03 09:10:50.475          2 Ave & E 96 St          7338.02   
4  2022-01-22 14:34:57.474          6 Ave & W 34 St          6364.10   

              end_station_name end_station_id  start_lat  start_lng  \
0  Mt Morris Park W & W 120 St        7685.14  40.802117 -73.968181   
1      Boerum Pl\t& Pacific St        4488.09  40.673746 -73.985649   
2              5 Ave & E 29 St        6248.06  40.761227 -73.960940   
3              5 Ave & E 29 St        6248.06  40.783964 -73.947167   
4              5 Ave & E 29 St        6248.06  40.749640 -73.988050   

     end_lat    end_lng member_casual  
0  40.804038 -73.945925        member  
1  40.688489 -73.991160        member  
2  40.745168 -73.986831        member  
3  40.745168 -73.986831        member  
4  40.745168 -73.986831        member

In [7]:
df.tail()

ride_id  rideable_type               started_at  \
29838801  1F223EDAFF420AE3  electric_bike  2022-12-01 20:26:45.847   
29838802  CFA5C560ACB73B8E   classic_bike  2022-12-26 13:46:34.237   
29838803  11C8C5E0DB947B07   classic_bike  2022-12-01 05:56:14.903   
29838804  5B9B083C534A5964   classic_bike  2022-12-02 11:54:15.871   
29838805  91C286C462F89A50   classic_bike  2022-12-18 13:35:22.574   

                         ended_at        start_station_name start_station_id  \
29838801  2022-12-01 20:30:46.012         Avenue D & E 3 St          5436.09   
29838802  2022-12-26 13:52:43.900            43 Ave & 47 St          6209.05   
29838803  2022-12-01 06:06:10.357         Avenue D & E 3 St          5436.09   
29838804  2022-12-02 12:01:00.747  Montague St & Clinton St          4677.06   
29838805  2022-12-18 13:37:27.193  Montague St & Clinton St          4677.06   

                     end_station_name end_station_id  start_lat  start_lng  \
29838801     Stanton St & Chrystie St        5523.02  40.720701 -73.977939   
29838802               39 Ave & 45 St        6401.03  40.744806 -73.917290   
29838803      Bleecker St & Crosby St        5679.08  40.720828 -73.977932   
29838804            Sands St & Jay St        4821.03  40.694271 -73.992327   
29838805  Cadman Plaza E & Tillary St        4677.01  40.694271 -73.992327   

            end_lat    end_lng member_casual  
29838801  40.722293 -73.991475        member  
29838802  40.749478 -73.918265        member  
29838803  40.726156 -73.995102        member  
29838804  40.700119 -73.986200        member  
29838805  40.695977 -73.990149        member

#### Joining through concatenate and making sure columns 5 and 7 were read as strings.   All files had the same number of headers (13) and there is ride_id as unique identifier.   

In [8]:
# List unique rideable types
df['rideable_type'].unique()

array(['electric_bike', 'classic_bike'], dtype=object)

In [9]:
df.isna().sum().sum()

np.int64(215066)

In [10]:
df.isna().sum().sort_values(ascending=False)

end_station_name      70092
end_station_id        70092
end_lng               37392
end_lat               37392
start_station_name       49
start_station_id         49
started_at                0
rideable_type             0
ride_id                   0
ended_at                  0
start_lat                 0
start_lng                 0
member_casual             0
dtype: int64

In [11]:
df[df.isna().any(axis=1)].head()


ride_id  rideable_type               started_at  \
170030  74EFAD3CDF3FCAD0  electric_bike  2022-01-12 05:34:03.133   
170031  C7D7571348DC1998  electric_bike  2022-01-18 22:47:32.232   
170033  71BFE3EF048771B0  electric_bike  2022-01-02 09:21:14.492   
170036  1666143231C2DD1F   classic_bike  2022-01-02 15:08:02.631   
170039  17EECF976DB9A86F  electric_bike  2022-01-18 17:23:46.723   

                       ended_at      start_station_name start_station_id  \
170030  2022-01-13 06:33:57.473         6 Ave & W 34 St          6364.10   
170031  2022-01-19 23:47:11.390         2 Ave & E 96 St          7338.02   
170033  2022-01-03 10:21:05.927         2 Ave & E 96 St          7338.02   
170036  2022-01-03 16:07:46.301  Riverside Dr & W 91 St          7524.16   
170039  2022-01-19 18:23:37.597  Riverside Dr & W 91 St          7524.16   

       end_station_name end_station_id  start_lat  start_lng  end_lat  \
170030              NaN            NaN  40.749640 -73.988050      NaN   
170031              NaN            NaN  40.783964 -73.947167      NaN   
170033              NaN            NaN  40.783964 -73.947167      NaN   
170036              NaN            NaN  40.793135 -73.977004      NaN   
170039              NaN            NaN  40.793135 -73.977004      NaN   

        end_lng member_casual  
170030      NaN        member  
170031      NaN        member  
170033      NaN        member  
170036      NaN        member  
170039      NaN        member

### There are between 37,392 and 70,092 trips with missing data: ending station/lat/lng.  May possibly recover the station name and id of those that do have a lat/lng, however, at least 32,700 records will be eliminated.

In [12]:
df.duplicated().sum()

np.int64(0)

## Get weather data using NOAA's API

## Load csv weather file sent via email 

In [13]:
# Replace with your actual file path
file_path = r"C:\Users\valev\CityBike\NOAA\4099324.csv"

# Only load 'start_time', 'end_time', 'start_station', 'end_station', 'duration'
cols_to_read = ['DATE', 'TAVG', 'PRCP']

df_noaa = pd.read_csv(file_path, usecols=cols_to_read)

# Quick check
print(df_noaa.head())       # first 5 rows
print(df_noaa.info())       # column info, dtypes, non-null counts


         DATE  PRCP  TAVG
0  2022-01-01  0.76    53
1  2022-01-02  0.04    53
2  2022-01-03  0.00    35
3  2022-01-04  0.00    27
4  2022-01-05  0.24    38
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   DATE    365 non-null    object 
 1   PRCP    365 non-null    float64
 2   TAVG    365 non-null    int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 8.7+ KB
None


In [14]:
df_noaa.isna().sum()

DATE    0
PRCP    0
TAVG    0
dtype: int64

In [15]:
df_noaa.shape

(365, 3)

In [16]:
df.dtypes

ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object

In [17]:
df_noaa.dtypes

DATE     object
PRCP    float64
TAVG      int64
dtype: object

In [18]:
# Drop rows where 'end_station_id' is NaN
df_clean = df.dropna(subset=['end_station_id'])

# Optional: reset the index
df_clean.reset_index(drop=True, inplace=True)

In [19]:
df_clean.isna().sum()

ride_id               0
rideable_type         0
started_at            0
ended_at              0
start_station_name    0
start_station_id      0
end_station_name      0
end_station_id        0
start_lat             0
start_lng             0
end_lat               0
end_lng               0
member_casual         0
dtype: int64

In [20]:

# Ensure datetime types
df_clean.loc[:, 'started_at'] = pd.to_datetime(df_clean['started_at'])


In [21]:
print(df_clean['started_at'].dtype)

object


In [24]:
# Extract only the date (drop time)
df_clean.loc[:, 'date_only'] = pd.to_datetime(df_clean['started_at']).dt.date
df_noaa.loc[:, 'date_only'] = pd.to_datetime(df_noaa['DATE']).dt.date

In [25]:
%%time
# Merge df_noaa into df, keeping only PRCP and TAVG
df_merge = df_clean.merge(
    df_noaa[['date_only', 'PRCP', 'TAVG']],
    on='date_only',
    how='left', indicator=True
)

# Optional: drop the helper column
df_merge.drop(columns=['date_only'], inplace=True)


CPU times: total: 8min 17s
Wall time: 14min 22s


In [26]:
df_merge.head()

ride_id  rideable_type                  started_at  \
0  BFD29218AB271154  electric_bike  2022-01-21 13:13:43.392000   
1  7C953F2FD7BE1302   classic_bike  2022-01-10 11:30:54.162000   
2  95893ABD40CED4B8  electric_bike  2022-01-26 10:52:43.096000   
3  F853B50772137378   classic_bike  2022-01-03 08:35:48.247000   
4  7590ADF834797B4B   classic_bike  2022-01-22 14:14:23.043000   

                  ended_at       start_station_name start_station_id  \
0  2022-01-21 13:22:31.463  West End Ave & W 107 St          7650.05   
1  2022-01-10 11:41:43.422             4 Ave & 3 St          4028.04   
2  2022-01-26 11:06:35.227          1 Ave & E 62 St          6753.08   
3  2022-01-03 09:10:50.475          2 Ave & E 96 St          7338.02   
4  2022-01-22 14:34:57.474          6 Ave & W 34 St          6364.10   

              end_station_name end_station_id  start_lat  start_lng  \
0  Mt Morris Park W & W 120 St        7685.14  40.802117 -73.968181   
1      Boerum Pl\t& Pacific St        4488.09  40.673746 -73.985649   
2              5 Ave & E 29 St        6248.06  40.761227 -73.960940   
3              5 Ave & E 29 St        6248.06  40.783964 -73.947167   
4              5 Ave & E 29 St        6248.06  40.749640 -73.988050   

     end_lat    end_lng member_casual  PRCP  TAVG _merge  
0  40.804038 -73.945925        member   0.0  21.0   both  
1  40.688489 -73.991160        member   0.0  35.0   both  
2  40.745168 -73.986831        member   0.0  28.0   both  
3  40.745168 -73.986831        member   0.0  35.0   both  
4  40.745168 -73.986831        member   0.0  21.0   both

In [27]:
df_merge.isna().sum()

ride_id                 0
rideable_type           0
started_at              0
ended_at                0
start_station_name      0
start_station_id        0
end_station_name        0
end_station_id          0
start_lat               0
start_lng               0
end_lat                 0
end_lng                 0
member_casual           0
PRCP                  432
TAVG                  432
_merge                  0
dtype: int64

In [29]:
df_merge['_merge'].value_counts(dropna = False)

_merge
both          29768282
left_only          432
right_only           0
Name: count, dtype: int64

In [35]:
# Show rows where weather data is missing
missing_weather = df_merge[df_merge['PRCP'].isna() | df_merge['TAVG'].isna()]

print(missing_weather.shape)   # confirm it's 432 rows
missing_weather.head(10)       # preview first 10


(432, 16)


ride_id  rideable_type                  started_at  \
4967   9D2DEF9B6D4FE1F2  electric_bike  2021-12-31 23:56:04.967000   
8335   07F3FEAE54546CCC   classic_bike  2021-12-31 23:25:37.862000   
9102   F7054D918A815DED  electric_bike  2021-12-31 23:38:49.704000   
9178   302DA77D1C2677A4  electric_bike  2021-12-31 23:41:16.311000   
18361  7A6CD4B02003DA01   classic_bike  2021-12-31 23:45:39.596000   
19138  077351EC0491BE6F   classic_bike  2021-12-31 23:21:56.609000   
21988  AC7AF19AC53F3DEE   classic_bike  2021-12-31 23:53:33.397000   
27208  E1AF0888109150B3   classic_bike  2021-12-31 23:30:44.237000   
29748  B54F016F64BC28D6   classic_bike  2021-12-31 23:28:34.147000   
30873  D958B82CA77BC0FF  electric_bike  2021-12-31 23:46:23.647000   

                      ended_at         start_station_name start_station_id  \
4967   2022-01-01 00:10:03.697  3 St & Prospect Park West          3865.05   
8335   2022-01-01 00:40:26.455        Hudson St & W 13 St          6115.06   
9102   2022-01-01 00:29:24.363            E 85 St & 3 Ave          7212.05   
9178   2022-01-01 00:03:43.417   Perry St & Greenwich Ave          5955.12   
18361  2022-01-01 00:09:44.400     W 84 St & Columbus Ave          7382.04   
19138  2022-01-01 00:39:39.163    Milton St & Franklin St          5752.07   
21988  2022-01-01 00:04:44.031         Broadway & E 14 St          5905.12   
27208  2022-01-01 00:04:03.570        Henry St & Grand St          5294.04   
29748  2022-01-01 00:05:32.632         Broadway & E 14 St          5905.12   
30873  2022-01-01 00:06:40.675           W 34 St & 11 Ave          6578.01   

               end_station_name end_station_id  start_lat  start_lng  \
4967   Atlantic Ave & Furman St        4614.04  40.668132 -73.973638   
8335            W 15 St & 6 Ave        5989.02  40.740057 -74.005274   
9102            E 81 St & 2 Ave        7121.05  40.778012 -73.954071   
9178            E 48 St & 3 Ave        6541.03  40.735918 -74.000939   
18361  Lexington Ave & E 120 St        7652.04  40.785000 -73.972834   
19138    Huron St & Franklin St        5869.04  40.729060 -73.957790   
21988           6 Ave & W 34 St        6364.10  40.734546 -73.990741   
27208       Henry St & Grand St        5294.04  40.714211 -73.981095   
29748           6 Ave & W 34 St        6364.10  40.734546 -73.990741   
30873           E 33 St & 1 Ave        6197.08  40.755942 -74.002116   

         end_lat    end_lng member_casual  PRCP  TAVG     _merge  
4967   40.691652 -73.999979        member   NaN   NaN  left_only  
8335   40.738046 -73.996430        casual   NaN   NaN  left_only  
9102   40.774779 -73.954275        member   NaN   NaN  left_only  
9178   40.754601 -73.971879        casual   NaN   NaN  left_only  
18361  40.801307 -73.939817        casual   NaN   NaN  left_only  
19138  40.732660 -73.958260        casual   NaN   NaN  left_only  
21988  40.749640 -73.988050        member   NaN   NaN  left_only  
27208  40.714211 -73.981095        casual   NaN   NaN  left_only  
29748  40.749640 -73.988050        casual   NaN   NaN  left_only  
30873  40.743227 -73.974498        casual   NaN   NaN  left_only

In [36]:
missing_weather.tail(10)

ride_id rideable_type                  started_at  \
9942537   7701DD7DD4322AE8  classic_bike  2021-11-13 17:52:37.376000   
10806931  8FC77EAE2C0561E4  classic_bike  2021-03-11 22:08:54.461000   
10977626  C86D57932E662AEB  classic_bike  2021-11-18 12:11:07.431000   
16722623  FA436228F4CD7647  classic_bike  2021-02-15 14:11:20.946000   
19549302  BA13AC6CD2C65E88  classic_bike  2021-07-10 14:53:21.751000   
21722627  763D82F8CFA1D9D1  classic_bike  2021-03-31 15:35:20.070000   
26940288  799622522BCC91B3  classic_bike  2021-09-05 17:52:23.693000   
27098850  55DE96B9AF7BFDF6  classic_bike  2021-07-19 14:41:06.067000   
28445458  87D33069747BE58B  classic_bike  2021-10-27 03:55:34.375000   
29325267  90E27C47E17E0F9B  classic_bike  2021-07-16 17:50:59.271000   

                         ended_at          start_station_name  \
9942537   2022-06-19 21:04:29.111          Broadway & W 25 St   
10806931  2022-06-16 09:12:21.565            Union St & 4 Ave   
10977626  2022-06-08 00:32:43.446        Carroll St & Bond St   
16722623  2022-08-10 15:01:30.839          Kent Ave & S 11 St   
19549302  2022-09-09 21:21:15.210    E Tremont Ave & E 176 St   
21722627  2022-09-25 14:33:21.607    Willoughby St & Fleet St   
26940288  2022-11-16 06:18:23.556     Anderson Ave & W 164 St   
27098850  2022-11-12 13:25:43.737        Fulton St & Pearl St   
28445458  2022-12-13 04:25:16.144     Greene Ave & Throop Ave   
29325267  2022-12-14 13:05:18.107  Bergen St & Vanderbilt Ave   

         start_station_id            end_station_name end_station_id  \
9942537           6173.08    Fairview Ave & Linden St        5066.03   
10806931          4175.15    Flatbush Ave & Ocean Ave        3704.04   
10977626          4184.07  E Fordham Rd & Webster Ave        8582.09   
16722623          5062.01      Union Ave & Jackson St        5300.06   
19549302          8356.02       Popham Ave & W 174 St        8348.02   
21722627          4628.05   DeKalb Ave & Franklin Ave        4528.01   
26940288          8096.03        Pier 40 Dock Station         SYS035   
27098850          5024.09            E 102 St & 1 Ave        7407.13   
28445458          4510.04     Ashland Pl & Dekalb Ave        4513.09   
29325267          4157.10        Morgan Loading Docks         SYS038   

          start_lat  start_lng    end_lat    end_lng member_casual  PRCP  \
9942537   40.742869 -73.989186  40.706940 -73.906320        casual   NaN   
10806931  40.677274 -73.982820  40.663657 -73.963014        casual   NaN   
10977626  40.678612 -73.990373  40.861748 -73.891050        casual   NaN   
16722623  40.707645 -73.968415  40.716075 -73.952029        casual   NaN   
19549302  40.847798 -73.901883  40.847746 -73.922079        casual   NaN   
21722627  40.691966 -73.981302  40.690648 -73.957462        casual   NaN   
26940288  40.832057 -73.927421  40.728660 -74.011980        casual   NaN   
27098850  40.707722 -74.004386  40.786995 -73.941648        casual   NaN   
28445458  40.689493 -73.942061  40.690065 -73.978776        casual   NaN   
29325267  40.679439 -73.968044  40.709306 -73.931175        casual   NaN   

          TAVG     _merge  
9942537    NaN  left_only  
10806931   NaN  left_only  
10977626   NaN  left_only  
16722623   NaN  left_only  
19549302   NaN  left_only  
21722627   NaN  left_only  
26940288   NaN  left_only  
27098850   NaN  left_only  
28445458   NaN  left_only  
29325267   NaN  left_only

### The missing values belong to the year 2021, therefore they will be eliminated. 

In [37]:
# Drop rows with missing weather
df_merge_clean = df_merge.dropna(subset=['PRCP', 'TAVG'])

print(df_merge_clean.shape)  # confirm row count reduced


(29768282, 16)


In [40]:
df_merge_clean.to_csv(r"C:\Users\valev\CityBike\nyc_data.csv", index=False)

In [39]:
import os
print(os.getcwd())

C:\Users\valev


In [1]:
df_merge_clean.columns

NameError: name 'df_merge_clean' is not defined